File: VociCore/Tests/VociCoreTests/Fixtures.swift
import Foundation
import VociCore

// MARK: - Deterministic time fixtures
//
// All tests anchor on a single fixed reference "now" and a fixed Gregorian calendar with an
// explicit UTC time zone, so "today"/"overdue" classification (and any date arithmetic) never
// depends on the machine's local timezone or the wall-clock time the tests happen to run at.

/// Fixed calendar used by every test: Gregorian, explicit UTC time zone.
let testCalendar: Calendar = {
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = TimeZone(identifier: "UTC")!
    return calendar
}()

/// Fixed reference instant: 2026-07-15 12:00:00 UTC.
let referenceNow: Date = {
    var components = DateComponents()
    components.calendar = testCalendar
    components.timeZone = testCalendar.timeZone
    components.year = 2026
    components.month = 7
    components.day = 15
    components.hour = 12
    components.minute = 0
    components.second = 0
    return testCalendar.date(from: components)!
}()

/// A deadline instant later on the same calendar day as `referenceNow` (still "today").
func todayLater(hour: Int = 18) -> Date {
    var components = testCalendar.dateComponents([.year, .month, .day], from: referenceNow)
    components.hour = hour
    return testCalendar.date(from: components)!
}

/// A deadline instant earlier on the same calendar day as `referenceNow` (still "today", and
/// also strictly before `referenceNow`, i.e. simultaneously "overdue" and "today" — both
/// classify as near-term).
func todayEarlier(hour: Int = 6) -> Date {
    var components = testCalendar.dateComponents([.year, .month, .day], from: referenceNow)
    components.hour = hour
    return testCalendar.date(from: components)!
}

/// A deadline `days` calendar days offset from `referenceNow`, preserving time-of-day.
/// Negative `days` yields a past (overdue) deadline; positive yields a future deadline.
func daysFromReferenceNow(_ days: Int) -> Date {
    testCalendar.date(byAdding: .day, value: days, to: referenceNow)!
}

/// A clearly-overdue deadline (yesterday, same time of day as `referenceNow`).
func yesterday() -> Date {
    daysFromReferenceNow(-1)
}

/// A clearly-future deadline beyond today (tomorrow, same time of day as `referenceNow`).
func tomorrow() -> Date {
    daysFromReferenceNow(1)
}

// MARK: - Deterministic id fixtures

/// Produces a stable, orderable `UUID` from a small integer so tests can reason about the
/// id-lexical tiebreak (tier 5) predictably: `fixedUUID(1)` sorts before `fixedUUID(2)`, etc.,
/// because both render as the same-length zero-padded hex suffix.
func fixedUUID(_ n: Int) -> UUID {
    let hex = String(format: "%012x", n)
    return UUID(uuidString: "00000000-0000-0000-0000-\(hex)")!
}

// MARK: - Task builder

/// Builds a `Task` with sensible defaults so each test only needs to specify the fields it
/// actually cares about.
func makeTask(
    id: UUID = UUID(),
    title: String = "Untitled task",
    status: TaskStatus = .todo,
    priority: Int? = nil,
    deadline: Date? = nil,
    conditions: [Condition] = [],
    estimateMinutes: Int? = nil,
    parentId: UUID? = nil,
    createdAt: Date = referenceNow
) -> Task {
    Task(
        id: id,
        title: title,
        status: status,
        priority: priority,
        deadline: deadline,
        conditions: conditions,
        estimateMinutes: estimateMinutes,
        parentId: parentId,
        createdAt: createdAt
    )
}

File: windows/tests/Volar.Core.Tests/Fixtures.cs
using System.Globalization;
using Volar.Core;

namespace Volar.Core.Tests;

/// <summary>
/// Deterministic time/id/task fixtures shared by every test in this project. Port of Swift's
/// <c>Fixtures.swift</c>.
/// </summary>
/// <remarks>
/// All tests anchor on a single fixed reference "now" and a fixed UTC time zone, so
/// "today"/"overdue" classification (and any date arithmetic) never depends on the machine's
/// local time zone or the wall-clock time the tests happen to run at.
/// </remarks>
internal static class Fixtures
{
    /// <summary>Fixed time zone used by every test: explicit UTC, never <see cref="TimeZoneInfo.Local"/>.</summary>
    public static readonly TimeZoneInfo TestTimeZone = TimeZoneInfo.Utc;

    /// <summary>Fixed reference instant: 2026-07-15 12:00:00 UTC.</summary>
    public static readonly DateTimeOffset ReferenceNow = new(2026, 7, 15, 12, 0, 0, TimeSpan.Zero);

    /// <summary>A deadline instant later on the same calendar day as <see cref="ReferenceNow"/> (still "today").</summary>
    public static DateTimeOffset TodayLater(int hour = 18) =>
        new(ReferenceNow.Year, ReferenceNow.Month, ReferenceNow.Day, hour, 0, 0, TimeSpan.Zero);

    /// <summary>
    /// A deadline instant earlier on the same calendar day as <see cref="ReferenceNow"/> (still
    /// "today", and also strictly before <see cref="ReferenceNow"/>, i.e. simultaneously "overdue"
    /// and "today" — both classify as near-term).
    /// </summary>
    public static DateTimeOffset TodayEarlier(int hour = 6) =>
        new(ReferenceNow.Year, ReferenceNow.Month, ReferenceNow.Day, hour, 0, 0, TimeSpan.Zero);

    /// <summary>
    /// A deadline <paramref name="days"/> offset from <see cref="ReferenceNow"/>, preserving
    /// time-of-day. Negative <paramref name="days"/> yields a past (overdue) deadline; positive
    /// yields a future deadline.
    /// </summary>
    public static DateTimeOffset DaysFromReferenceNow(int days) => ReferenceNow.AddDays(days);

    /// <summary>A clearly-overdue deadline (yesterday, same time of day as <see cref="ReferenceNow"/>).</summary>
    public static DateTimeOffset Yesterday() => DaysFromReferenceNow(-1);

    /// <summary>A clearly-future deadline beyond today (tomorrow, same time of day as <see cref="ReferenceNow"/>).</summary>
    public static DateTimeOffset Tomorrow() => DaysFromReferenceNow(1);

    /// <summary>
    /// Produces a stable, orderable <see cref="Guid"/> from a small integer so tests can reason
    /// about the id-ordinal tiebreak (tier 5) predictably: <c>FixedGuid(1)</c> sorts before
    /// <c>FixedGuid(2)</c>, etc., because both render as the same-length zero-padded hex suffix.
    /// </summary>
    public static Guid FixedGuid(int n)
    {
        var hex = n.ToString("x12", CultureInfo.InvariantCulture);
        return Guid.Parse($"00000000-0000-0000-0000-{hex}", CultureInfo.InvariantCulture);
    }

    /// <summary>
    /// Builds a <see cref="TaskSnapshot"/> with sensible defaults so each test only needs to
    /// specify the fields it actually cares about. Port of Swift's <c>makeTask(...)</c> free
    /// function.
    /// </summary>
    public static TaskSnapshot MakeTask(
        Guid? id = null,
        string title = "Untitled task",
        TaskState status = TaskState.Todo,
        int? priority = null,
        DateTimeOffset? deadline = null,
        IReadOnlyList<Condition>? conditions = null,
        int? estimateMinutes = null,
        Guid? parentId = null,
        DateTimeOffset? createdAt = null) =>
        new(
            id ?? Guid.NewGuid(),
            title,
            status,
            priority,
            deadline,
            conditions ?? Array.Empty<Condition>(),
            estimateMinutes,
            parentId,
            createdAt ?? ReferenceNow);
}
